Co-Existing with Boto

eden-ohana opened the pull request:
Fixes #2624

Supports wildcard matching on prefix bucket (With known bug on list-objects)

TO DO- readme and tests

eden-ohana committed:
botwo.py

eden-ohana committed:
botwo.py

arielshaqed commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+
+class Paginatwo(object):
+
+    def __init__(self, mapping, kwargs, client1, client2, client1_name, client2_name):
```
Not sure of this usage of "`kwargs`".  AFAIU this Python name is explicitly used for `**kwargs`.  Perhaps just `args` here?

eden-ohana replied:
Done

arielshaqed replied:
this is not done, we still have `kwargs` in a parameter list of a method.

eden-ohana committed:
update README.md

johnnyaug commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+import fnmatch
+
+SPECIAL_METHODS = {"get_paginator", "delete_objects"}
```
"SPECIAL" how? be more specific

eden-ohana replied:
Done

johnnyaug commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+    return botwo.create_client_class(client_mapping, config)
+
+
+class Paginatwo(object):
```
Please rename this to something meaningful.
Also add documentation

eden-ohana replied:
Done

arielshaqed commented on README.md:
````diff
@@ -1 +1,143 @@
+```
+
+## Configuration
+* client_mapping(dict) - The mapping of the two boto client names for the config. For example, ```{"s3": s3, "minio": minio}```
````
Is there anything in the design or the code that limits us to two clients?

arielshaqed commented on README.md:
```diff
@@ -1 +1,143 @@
+    { 
+        "bucket_name": "example-bucket", # required - the bucket name or prefix to route from, supports wildcard matching (example-bucket/prefix/*)
+        "mapped_bucket_name": "new-bucket", # optional - the new bucket name to route from bucket name
+        "client": "minio", # optinal - the client to call with this bucket name, if not provided will go to default
```
```suggestion
        "client": "minio", # optional - the client to call with this bucket name, if not provided will go to default
```

arielshaqed commented on README.md:
```diff
@@ -1 +1,143 @@
+    },   
+    {
+        "bucket_name": "*", 
+        "client": "s3", # default client - required
```
I don't understand.  Line 50 said that there is a default client _without_  having to configure this field.

arielshaqed commented on README.md:
```diff
@@ -1 +1,143 @@
+        "key_prefix": "test/" # optional - add prefix to new files inside the bucket (put_object(Bucket="example-bucket", "Key"="obj.py") --> new-bucket/test/obj.py)
+    },   
+    {
+        "bucket_name": "*", 
```
Please document what asterisks do.  I can think of any of these interpretations:

* Match _any_ substring (like an asterisk in the shell)?  (Sometimes called a "wildcard")
* Match any suffix, and are legal only once and at the end of a string?
* Only the string `"*"` is allowed, which matches "anything"?

arielshaqed commented on README.md:
```diff
@@ -1 +1,143 @@
+config = [
+    { 
+        "bucket_name": "example-bucket", # required - the bucket name or prefix to route from, supports wildcard matching (example-bucket/prefix/*)
+        "mapped_bucket_name": "new-bucket", # optional - the new bucket name to route from bucket name
```
```suggestion
        "mapped_bucket_name": "new-bucket", # optional - the new bucket name to use; bucket name unchanged if not specified.
```

arielshaqed commented on README.md:
```diff
@@ -1 +1,143 @@
+import boto3
+import botwo
+
+# Initializing two boto s3 clients according to boto API
```
```suggestion
# Initialize two boto s3 clients according to boto API
```

Every docstring in [PEP-0257](https://www.python.org/dev/peps/pep-0257/) is an imperative (like this one) rather than a perfect.  And most style guides recommend this style.

(Also comments below)

arielshaqed commented on README.md:
````diff
@@ -1 +1,143 @@
+## Using Botwo with [lakeFS]
+When a user with an existing python code that uses boto3 client to access s3 wants to move only part of their buckets/prefixes to lakefs, botwo allows using lakefs and s3 together with minimum code changes.
+
+For example: user can route all requests to s3 with ```bucket-a``` to lakeFs with ```example-repo/main/``` in lakeFS (example-repo = repository name).
````
```suggestion
For example: route all requests to s3 with ```bucket-a``` to lakeFs with ```example-repo/main/``` in lakeFS (example-repo = repository name).
```

(Saying "user can" is more appropriate for requirements than for users.  Users just want to know what to do, they already know they are users.)

Also try to avoid "user" in the rest of the docs.

arielshaqed commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+        self.client1_name = client1_name
+        self.client2_name = client2_name
+        self.paginator1 = client1.get_paginator(kwargs["operation_name"])
+        self.paginator2 = client2.get_paginator(kwargs["operation_name"])
```
Not sure why we need to change from the input dicts defined above to this.  It just limits us to 2 clients, and adds many `if`s to the code.  I think we would be better to hold the 2 dicts directly.

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+                    else:
+                        raise ValueError("Invalid client name " + map.get("client"))
+                else:
+                    raise ValueError("client isn't provided")
```
This error should be thrown once during construction -- the input parameters are malformed -- rather than at runtime when we hit a row in the configuration that has no "client" field.

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+                    elif map.get("client") == self.client1_name:
+                        paginator_to_call = self.paginator1
+                    else:
+                        raise ValueError("Invalid client name " + map.get("client"))
```
E.g. if we used the user-supplied dicts directly this would be:
```suggestion
                try:
                    client = self.clients_by_name[map["client"]]
                except KeyError:
                    raise UnknownClientError(map["client"])
```
(I'm assuming we've already validated all map rows during construction, as suggested in the next comment.)

eden-ohana replied:
Done

johnnyaug commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+        class_attributes = self._create_methods()
+
+        bases = [Botwo]
+        cls = type("s3", tuple(bases), class_attributes)
```
Not sure why the base of Botwo is needed here. If it's for the "meta" member, you can explicitly add it as an attribute.

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+                    else:
+                        raise ValueError("Invalid client name " + map.get("client"))
+                else:
+                    raise ValueError("client isn't provided")
```
Also: wherever we throw this error, it **must** say which row in the configuration is bad.  Otherwise how do I debug this error when I get it?

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+                    kwargs["Bucket"] = map.get("mapped_bucket_name")
+                if kwargs.get("Prefix"):
+                    if map.get("key_prefix"):
+                        kwargs["Prefix"] = map.get("keu_prefix") + kwargs["Prefix"]
```
How was this tested?  I believe there is an error on this line.
```suggestion
                        kwargs["Prefix"] = map.get("key_prefix") + kwargs["Prefix"]
```

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+        if type(client_mapping) is not dict:
+            raise TypeError("Invalid client mapping type: " + str(type(client_mapping)) + " expected dict")
+        if len(client_mapping) != 2:
+            raise TypeError("two clients is expected")
```
```suggestion
            raise TypeError("two clients expected")
```
or
```suggestion
            raise TypeError("two clients are expected")
```
(or replace "expected" with "needed" or "required")

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+        if len(client_mapping) != 2:
+            raise TypeError("two clients is expected")
+
+        names = list(client_mapping)
```
I don't understand how this works.  I could not find anywhere in the Python documentation that `list(DICT)` returns the keys of DICT in any particular order.  So two runs of the same code might set `client1` and `client2` in opposite ways.  This will make debugging less fun.

arielshaqed commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+    return botwo.create_client_class(client_mapping, config)
+
+
+class Paginatwo(object):
```
According to [PEP-8](https://www.python.org/dev/peps/pep-0008/#documentation-strings), this class and every user-visible method should have a docstring.

eden-ohana replied:
Done

johnnyaug commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+        self.mapping = None
+        self.meta = None
+
+    def create_client_class(self, client_mapping, config):
```
1. I think this method should return an _instance_ of the created class, and be called create_client.

2. This class IIUC - is more like a BotwoBuilder, and this method should be called _build()_.

eden-ohana replied:
Done

johnnyaug replied:
create_client_class does not create a class anymore - can be renamed to "build"

eden-ohana replied:
Done

johnnyaug commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+    def create_client_class(self, client_mapping, config):
+        if type(client_mapping) is not dict:
+            raise TypeError("Invalid client mapping type: " + str(type(client_mapping)) + " expected dict")
+        if len(client_mapping) != 2:
```
Why?
Can't we have an unlimited number?

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+        self.mapping = None
+        self.meta = None
+
+    def create_client_class(self, client_mapping, config):
```
Can we verify here that all the clients are actually Boto S3 clients?  Otherwise we will generate a _weird_ class, and the user will see very strange errors...

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+        class_attributes = self._create_methods()
+
+        bases = [Botwo]
+        cls = type("s3", tuple(bases), class_attributes)
```
IIUC this creates a new type that is named "s3".  That is confusing (at least to me), because this class is called `Botwo` and (I think...) it pretends to be an object of type [`S3.Client`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client).

johnnyaug commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+        def _delete_objects_api_call(*args, **kwargs):
+            if args:
+                raise TypeError("%s() only accepts keyword arguments." % operation_name)
+            # TODO
```
TODO

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+        def _delete_objects_api_call(*args, **kwargs):
+            if args:
+                raise TypeError("%s() only accepts keyword arguments." % operation_name)
+            # TODO
```
```suggestion
            # TODO(eden)
```
or
```suggestion
            # TODO(ISSUE NUMBER)
```

Can we raise an exception here?  Returning `None` makes the user think that the call succeeded.

eden-ohana replied:
I implemented it.

arielshaqed commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+        operations = [func for func in dir(self.default_client) if
+                      (callable(getattr(self.default_client, func)) and not func.startswith('_'))]
+        for operation_name in operations:
+            if operation_name in SPECIAL_METHODS:
+                op_dict[operation_name] = self._create_special_method(operation_name)
```
Since `_create_special_method` will _anyway_ run a switch on `operation_name`, why not just do that switch directly here?  I am not sure what the extra level of indirection helps us here.

eden-ohana replied:
Done

arielshaqed reviewed (changes requested):
Thanks!  This is some really elegant and useful code.

All the user-visible changes are blocking (at least in terms of discussing them): this is a client for users, so I really want it to be easy for users to have an easy time configuring things correctly (because we made it hard to make a mistake), and then to understand errors if they do get them.

Not a Python expert, so these may be built-in in ways I do not see or understand...

1. All user-visible objects should have docstrings,
1. Where do we run pylint?

johnnyaug commented on README.md:
```diff
@@ -1 +1,143 @@
+s3_west = boto3.client('s3', region_name='us-west-1')
+
+# Defining the rules for the routing between the clients
+config = [
+    {
+        "bucket_name": "bucket/a/*",
+        "mapped_bucket_name": "new-bucket",
+        "client": "s3_west", 
+        "key_prefix": "test/"
+    },
+    {
+        "bucket_name": "*",
+        "client": "s3_east", # default client
+    }
+]
+
+# Defining the mapping between the boto clients to the client in the config
+client_mapping = {"s3_east": s3_east, "s3_west": s3_west}
```
```suggestion
profiles = [
    "s3_west": {
        "source_bucket_pattern": "my_bucket",
        "source_key_pattern": "my_prefix/*",
        "mapped_bucket_name": "new-bucket",
        "mapped_prefix": "test/"
    },
    "s3_east": {
        "source_bucket_pattern": "my_bucket_prefix*",
        "mapped_bucket_name": "new-bucket",
        "mapped_prefix": "test2/"
    }
]

# Defining the mapping between the boto clients to the client in the config
client_mapping = {"s3_east": s3_east, "s3_west": s3_west, "default": s3_east}
```

johnnyaug commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+        for map in self.mapping:
+            if fnmatch.fnmatch(kwargs["Bucket"], map["bucket_name"]):
+                if map.get("client"):
+                    if map.get("client") == self.client2_name:
```
This should probably come from the map instead of comparing names

eden-ohana replied:
Done

johnnyaug commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+                    kwargs["Bucket"] = map.get("mapped_bucket_name")
+                if kwargs.get("Prefix"):
+                    if map.get("key_prefix"):
+                        kwargs["Prefix"] = map.get("keu_prefix") + kwargs["Prefix"]
```
```suggestion
                        kwargs["Prefix"] = map.get("key_prefix") + kwargs["Prefix"]
```

eden-ohana replied:
Done

johnnyaug commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+                    raise ValueError("client isn't provided")
+
+                if not is_copy:
+                    if map.get("mapped_bucket_name"):
```
constant for all literal string in the file

johnnyaug commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+    def _route_params(self, path, api_params, is_copy):
+        client_to_call = self.default_client
+        for map in self.mapping:
+            if fnmatch.fnmatch(path, map.get("bucket_name")):
+                if map.get("client"):
+                    if map.get("client") == self.client2_name:
+                        client_to_call = self.client2
+                    elif map.get("client") == self.client1_name:
+                        client_to_call = self.client1
+                    else:
+                        raise ValueError("Invalid client name " + map.get("client"))
+                else:
+                    raise ValueError("client isn't provided")
```
I think this logic is out of scope for _route_params?

eden-ohana replied:
Done

johnnyaug commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+    def _route_params(self, path, api_params, is_copy):
+        client_to_call = self.default_client
+        for map in self.mapping:
+            if fnmatch.fnmatch(path, map.get("bucket_name")):
+                if map.get("client"):
+                    if map.get("client") == self.client2_name:
+                        client_to_call = self.client2
+                    elif map.get("client") == self.client1_name:
+                        client_to_call = self.client1
+                    else:
+                        raise ValueError("Invalid client name " + map.get("client"))
+                else:
+                    raise ValueError("client isn't provided")
```
The case where the client isn't provided should have already failed when building the instance.

eden-ohana replied:
Done

johnnyaug commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+                    if map.get("key_prefix"):
+                        api_params["Key"] = map.get("key_prefix") + api_params["Key"]
+                else:
+                    source_path = api_params["CopySource"].split("/")
```
According to the boto documentation, CopySource is a dictionary.

eden-ohana replied:
Done

johnnyaug commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+                else:
+                    raise ValueError("client isn't provided")
+
+                if not is_copy:
```
I think this part should happen also for copy methods

johnnyaug commented on setup.py:
```diff
@@ -0,0 +1,34 @@
+
+"""
+
+from setuptools import setup, find_packages
```
Can we separate all publishing/installation related stuff to a different PR?

eden-ohana replied:
Done

johnnyaug commented on botwo/botwo.py:
```diff
@@ -0,0 +1,188 @@
+                raise TypeError("%s() only accepts keyword arguments." % operation_name)
+            client_to_call = self.default_client
+            if kwargs.get("Bucket"):  # bucket operation
+                full_path = kwargs["Bucket"] + "/" + kwargs["Key"]
```
IMO you should never concatenate bucket and key. It's a bug waiting to happen

johnnyaug replied:
Some operations operate on a `Prefix`, rather than a key (e.g. list objects operations). This will result in a KeyError.

eden-ohana replied:
Done

johnnyaug reviewed (changes requested):
Really awesome work!

eden-ohana committed:
update botwo.py

eden-ohana committed:
update botwo.py and test_client.py

eden-ohana committed:
update botwo.py

arielshaqed commented on botwo/__init__.py:
```diff
@@ -0,0 +1,14 @@
+
+def client(client_mapping, profiles):
+    """
+    Create a Botwo client that routes between boto clients by configuration.
```
This docstring should explain the parameters.  (It is the right place to explain parameters to users, that way it shows up in the Python REPL and on [PyPI](https://pypi.org/).

arielshaqed replied:
I understand we are leaving docstrings for the next PR.  If this still holds then you can resolve all "docstring" comments.

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -1,131 +1,144 @@
+
+
+class BotwoBuilder(object):
+    """ This class creates botwo client that wraps boto client:
```
```suggestion
    """This class creates a botwo client that wraps boto clients:
```

eden-ohana replied:
Done

arielshaqed commented on tests/test_client.py:
```diff
@@ -0,0 +1,188 @@
+                                                   environment={"MINIO_ROOT_USER": "minioadmin",
+                                                                "MINIO_ROOT_PASSWORD": "minioadmin",
+                                                                "MINIO_UPDATE": "off"},
+                                                   ports={"9000": 9000}, detach=True)
```
Need to allow Docker to pick an ephemeral port, and then use that.  Otherwise this test _will not run_ on a machine that is using ports 9000 or 7000.

eden-ohana replied:
Done

eden-ohana commented:
Thanks for the review :) I addressed your comments.

The readme, workflows and publish files will be added in a separate PR.
I opened issues with the missing functionalities.

@arielshaqed I will add pylint to the test workflow.

eden-ohana committed:
update botwo.py

arielshaqed commented on botwo/botwo.py:
```diff
@@ -1,131 +1,144 @@
-    botwo = Botwo()
-    return botwo.create_client_class(client_mapping, config)
+class PaginatorWrapper(object):
+    """ This class wraps boto paginator """
```
This docstring can be more helpful to the user, and also satisfy PEP-257.  E.g.:
```suggestion
    """Wrapper for a boto paginator.
    
    Holds multiple paginators, one for each client, and dispatches calls to the appropriate
    paginator according to ...
    """
```
("..." needs to be updated to describe what it does).

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -1,131 +1,144 @@
-class Paginatwo(object):
-
-    def __init__(self, mapping, kwargs, client1, client2, client1_name, client2_name):
+    def __init__(self, mapping, config, kwargs):
```
This is user-visible, so according to PEP-8 and PEP-257 should have a docstring.  This docstring is a good place to describe what `mapping` and `config` are.

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -1,131 +1,144 @@
-class Paginatwo(object):
-
-    def __init__(self, mapping, kwargs, client1, client2, client1_name, client2_name):
+    def __init__(self, mapping, config, kwargs):
```
I don't understand this usage of "`kwargs`".  I don't think I have ever seen it in an arg list as anything except "`**kwargs` -- when its usage becomes apparent.

arielshaqed replied:
This is still relevant.  "`kwargs`" used like this is confusing.

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -1,131 +1,144 @@
+        self.config = config
+        self.paginators = dict()
+        for client in self.mapping:
+            self.paginators[client] = self.mapping[client].get_paginator(kwargs["operation_name"])
```
If this function took *two* args `operation_name` and `**kwargs`, then it could simply _ignore_ `kwargs` and access `operation_name` directly.

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -1,131 +1,144 @@
-            raise TypeError("Invalid client mapping type: " + str(type(client_mapping)) + " expected dict")
-        if len(client_mapping) != 2:
-            raise TypeError("two clients is expected")
+    def build(self, mapping, config):
```
This function needs a docstring according to PEP-8 and PEP-257.  That docstring should document `config` and `mapping`.

eden-ohana replied:
Done

johnnyaug commented on botwo/botwo.py:
```diff
@@ -0,0 +1,200 @@
+        for profile in self.config:
+            mapping = self.config[profile]
+
+            if fnmatch.fnmatch(api_params["Bucket"], mapping.get("source_bucket_pattern")):
+                if api_params["Key"]:
+                    if mapping.get("source_key_pattern"):
+                        if not fnmatch.fnmatch(api_params["Key"], mapping.get("source_key_pattern")):
+                            continue
+                    if mapping.get("mapped_prefix"):
+                        api_params["Key"] = mapping.get("mapped_prefix") + api_params["Key"]
+                if mapping.get("mapped_bucket_name"):
+                    api_params["Bucket"] = mapping.get("mapped_bucket_name")
```
All this code is basically duplicated in the paginator.

This function can be `_route_bucket_and_key(bucket, key)`.
Then you can use it in the paginator as is.
Also then the usage in line 117 will be more natural instead of artificially creating a map to pass down to this function.

arielshaqed commented on botwo/botwo.py:
```diff
@@ -1,131 +1,144 @@
-        return config
+        for profile in self.config:
+            if not self.mapping.get(profile):
+                raise ValueError("profile: " + profile + " mapping is required")
```
```suggestion
                raise ValueError("profile " + profile + " in config does not appear in mapping")
```

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -137,52 +150,51 @@ def _api_call(*args, **kwargs):
-                    api_params["CopySource"] = source_bucket + "/" + source_key
-                break
-        return client_to_call, api_params
+            if kwargs.get("Delete"):  # delete objects operation
```
```suggestion
            if "Delete" in kwargs:  # delete objects operation
```

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -137,52 +150,51 @@ def _api_call(*args, **kwargs):
+            if kwargs.get("Delete"):  # delete objects operation
+                for i, obj in enumerate(kwargs["Delete"]["Objects"]):
+                    params = {"Bucket": kwargs.get("Bucket"), "Key": obj["Key"]}
+                    res = self._route_params(params)
+                    client_to_call = res[0]
+                    bucket = res[1]["Bucket"]
```
Can we do this instead?

```suggestion
                    client_to_call, b = self._route_params(params)
                    bucket = b["Bucket"]
```
(also in all calls to `_route_params`).

Also I think other places destructure like this -- and it is quite hard for me to read.

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -137,52 +150,51 @@ def _api_call(*args, **kwargs):
+                    client_to_call = res[0]
+                    bucket = res[1]["Bucket"]
+                    kwargs["Delete"]["Objects"][i]["Key"] = res[1]["Key"]
+                    # TODO (eden) fail if client to call not equal to previous
```
IIUC deletion will *fail* when that happens, so best mark it as a "BUG".

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -137,52 +150,51 @@ def _api_call(*args, **kwargs):
+                        if not fnmatch.fnmatch(api_params["Key"], mapping.get("source_key_pattern")):
+                            continue
+                    if mapping.get("mapped_prefix"):
+                        api_params["Key"] = mapping.get("mapped_prefix") + api_params["Key"]
```
I still don't understand the rules language for editing keys.  But IIUC it allows me to *add* a prefix at the start of the key, but not to *trim* a prefix.  Is there a particular reason for that?

arielshaqed reviewed (changes requested):
Thanks!

I still do not understand how users will understand how to configure the rules engine.  It is a dict of dicts.  So I think we should document its structure exactly.

arielshaqed commented on botwo/botwo.py:
```diff
@@ -0,0 +1,200 @@
+        for profile in self.config:
+            mapping = self.config[profile]
+
+            if fnmatch.fnmatch(kwargs["Bucket"], mapping.get("source_bucket_pattern")):
```
Again, if we declared `paginate(self, Bucket, **kwargs)`, would it work?  Would this become easier to read?

arielshaqed commented on botwo/botwo.py:
```diff
@@ -0,0 +1,200 @@
+        for profile in self.config:
+            mapping = self.config[profile]
+
+            if fnmatch.fnmatch(kwargs["Bucket"], mapping.get("source_bucket_pattern")):
```
If called without a `Bucket=` argument, this causes
```
KeyError: 'Bucket'
```
Is this a good error?

eden-ohana replied:
Done

arielshaqed commented on botwo/botwo.py:
```diff
@@ -0,0 +1,200 @@
+            mapping = self.config[profile]
+
+            if fnmatch.fnmatch(kwargs["Bucket"], mapping.get("source_bucket_pattern")):
+                if kwargs["Prefix"]:
+                    if mapping.get("source_key_pattern"):
+                        if not fnmatch.fnmatch(kwargs["Prefix"], mapping.get("source_key_pattern")):
+                            continue
+                    if mapping.get("mapped_prefix"):
+                        kwargs["Prefix"] = mapping.get("mapped_prefix") + kwargs["Prefix"]
```
AFAIU Python will give a `KeyError` here unless `kwargs` has a `Prefix` key.  So this only checks whether `Prefix` is empty, right?

So I think this is a bug: if `Prefix` is empty then it does not prepend `mapped_prefix`, and the prefix stays empty.  E.g. (I think) listing the empty prefix will be wrong if the default has a non-empty `mapped_prefix`.

I think this simpler code is better because it fixes the bug that I think we have:
```suggestion
                if mapping.get("source_key_pattern"):
                    if not fnmatch.fnmatch(kwargs["Prefix"], mapping.get("source_key_pattern")):
                        continue
                if mapping.get("mapped_prefix"):
                    kwargs["Prefix"] = mapping.get("mapped_prefix") + kwargs["Prefix"]
```

arielshaqed commented on botwo/botwo.py:
```diff
@@ -0,0 +1,200 @@
+            if args:
+                raise TypeError("%s() only accepts keyword arguments." % operation_name)
+            client_to_call = self.default
+            if kwargs.get("Bucket") and not kwargs.get("Prefix"):
```
Do we truly want to ignore empty `Prefix` (as opposed to calls where `Prefix` is not specified)? AFAIU this will do the wrong thing if I try to list with an empty `Prefix`.

This pattern of calling `kwargs.get(k)` and ignoring the return value instead of `k in kwargs` also appears in other places, and I am not sure it is correct.  (But I am not a Python expert, and will be happy to learn that I am incorrect in thinking that these are errors!)

arielshaqed commented on tests/test_client.py:
```diff
@@ -0,0 +1,188 @@
+        self.botwo_client.put_object(Bucket="test-list-objects", Key="1.txt", Body=content)
+        self.botwo_client.put_object(Bucket="test-list-objects", Key="test/3.txt", Body=content)
+
+        objects = self.botwo_client.list_objects(Bucket="list", Prefix="test")
```
Please also add a test with empty `Prefix`, i.e. where after setup you say
```python
        objects = self.botwo_client.list_objects(Bucket="list", Prefix="")
```

eden-ohana replied:
Done

arielshaqed commented on tests/test_client.py:
```diff
@@ -0,0 +1,188 @@
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.docker1.kill()
```
Can we `stop` instead?  I prefer to `docker stop`, the container might need a bit of notice to get all of its logs out, particularly if the process failed to start.

eden-ohana replied:
Done

arielshaqed reviewed (commented):
Some more comments, mostly about docker behaviour.

eden-ohana committed:
changes from PR

eden-ohana committed:
changes from PR

arielshaqed commented on botor/botor.py:
```diff
@@ -0,0 +1,211 @@
+            if fnmatch.fnmatch(api_params["Bucket"], mapping.get("source_bucket_pattern")):
+                if "Key" in api_params:
+                    if "source_key_pattern" in mapping:
+                        if not fnmatch.fnmatch(api_params["Key"], mapping.get("source_key_pattern")):
```
We mix 2 styles here for keys that we know are in the dict.  Here we have `if "source_key_pattern" in mapping: ... mapping.get("source_key_pattern")`, but `if "Key" in api_params: ... api_params["Key"]`.  Let's have one (and I prefer `[...]`, because that throws an exception if I made a mistake.

eden-ohana replied:
Done

arielshaqed replied:
I still see the old code; perhaps you missed a `git push` from your local repo?

arielshaqed commented on tests/test_client.py:
```diff
@@ -0,0 +1,197 @@
+        self.assertEqual(len(objects['Contents']), 2)
+        # test empty prefix
+        objects = self.botwo_client.list_objects(Bucket="list", Prefix="")
+        self.assertEqual(len(objects['Contents']), 3)
```
I think we need to test the return values here -- they may need to be edited back to the correct prefix; see https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html#API_ListObjectsV2_Example_3 for what responses look like.

This can be left as an open issue to do in a separate PR; just open the issue and mark # BUG(#123): ... to link the bug number.

eden-ohana replied:
Done

arielshaqed replied:
This remains open.

johnnyaug commented on botor/__init__.py:
```diff
@@ -0,0 +1,29 @@
+
+
+def client(client_mapping, profiles):
+    """Create a botor client that routes between boto clients by configuration.
```
```suggestion
    """Create a botor client that routes between boto3 s3 clients by configuration.
```

eden-ohana replied:
Done

arielshaqed commented on tests/test_client.py:
```diff
@@ -0,0 +1,197 @@
+                                                                "MINIO_UPDATE": "off"},
+                                                   ports={"9000": None}, detach=True)
+
+        cls.docker1.reload()
```
Are we guaranteed that we need an immediate `reload`, and that after it we are guaranteed to see the port and to have a working minio?

eden-ohana replied:
Yes

arielshaqed commented on botor/botor.py:
```diff
@@ -0,0 +1,211 @@
+        self.mapping = None
+        self.config = None
+
+    def build(self, mapping, config):
```
Any reason why I need to create the object and then call its `build` method?  Could we just have `__init__(self, mapping, config)` with this body?

eden-ohana replied:
Yes, we must use the build method in order to generate the boto methods on the fly. ``__init__(self, mapping, config)`` will not work

arielshaqed replied:
Oh, I see!  So this is not really a class, just a function.

I'll refactor it (in a separate) to remove this class entirely.

arielshaqed reviewed (dismissed):
Thanks!

I'm approving -- thanks for going through this long process!  All issues are either mechanical and can be handled (mixing `.get` and `[]` on dicts when we know the key is in the dict, or the empty constructor that requires calling `build`), or they can be handled by opening issues and just linking them into the code. :-)

johnnyaug commented on botor/__init__.py:
```diff
@@ -0,0 +1,29 @@
+                                                       ;supports wildcard matching (prefix/a/*))
+               "mapped_bucket_name": "new-bucket", (optional - the new bucket name to use;
+                                                   bucket name unchanged if not specified)
+               "mapped_prefix": "test/" (optional - add prefix to new files inside the bucket.
```
```suggestion
               "mapped_prefix": "test/" (optional - add this to the given key/prefix when routing the request to the mapped bucket.
```

eden-ohana replied:
Done

johnnyaug commented on botor/__init__.py:
```diff
@@ -0,0 +1,29 @@
+               For example,(put_object(Bucket="example-bucket", "Key"="a/obj.py") --> new-bucket/test/a/obj.py))
+           },}
+
+    :returns: botor client that maps 1:1 with boto API
```
```suggestion
    :returns: a botor client, compatible with the boto S3 client.
```

eden-ohana replied:
Done

johnnyaug reviewed (approved):
Thanks! Left documentation comments

arielshaqed commented on tests/test_client.py:
```diff
@@ -0,0 +1,197 @@
+        self.botwo_client.put_object(Bucket="test-list-objects", Key="1.txt", Body=content)
+        self.botwo_client.put_object(Bucket="test-list-objects", Key="test/3.txt", Body=content)
+
+        objects = self.botwo_client.list_objects(Bucket="list", Prefix="test")
```
*MUST* test this with [`list_objects_v2`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.list_objects_v2) -- this is what users should be using.

The documentation actually warns users *AGAINST* using ListObjects, e.g. (on that page):
> Warning
>
> This action has been revised. We recommend that you use the newer version, ListObjectsV2 , when developing applications. For backward compatibility, Amazon S3 continues to support ListObjects .

arielshaqed commented on tests/test_client.py:
```diff
@@ -0,0 +1,197 @@
+                params['Bucket'] = 'test-meta'
+
+        event_system.register('provide-client-params.s3.GetBucketAcl', add_my_bucket)
+        self.botwo_client.get_bucket_acl()
```
Need to test return value here, too.

arielshaqed commented on tests/test_client.py:
```diff
@@ -0,0 +1,197 @@
+        contents = data['Body'].read().decode('utf-8')
+        self.assertEqual(contents, content)
+
+        with pytest.raises(Exception):
```
This is Python!  Need to test a particular exception, otherwise even some syntax and logic errors in the called code (or in the test code) will make the test pass.

Also in every call to `pytest.raises(Exception)`

arielshaqed commented on tests/test_client.py:
```diff
@@ -0,0 +1,197 @@
+
+
+def create_botwo_client(minio1, minio2):
+    profiles = {
+        "test_upload": {
+            "source_bucket_pattern": "bucket-a",
+            "source_key_pattern": "a/*",
+            "mapped_bucket_name": "test-upload",
+            "mapped_prefix": "main/"
+        },
+        "test_copy": {
+            "source_bucket_pattern": "source",
+            "mapped_bucket_name": "test-copy",
+            "mapped_prefix": "main/"
+        },
+        "test_copy2": {
+            "source_bucket_pattern": "test-copy",
+        },
+        "test_delete_objects": {
+            "source_bucket_pattern": "delete-objects*",
+            "mapped_bucket_name": "test-delete-objects",
+            "mapped_prefix": "test/"
+        },
+        "test_paginator": {
+            "source_bucket_pattern": "pagi",
+            "mapped_bucket_name": "test-paginator",
+            "mapped_prefix": "dev/"
+        },
+        "test_list_objects": {
+            "source_bucket_pattern": "list",
+            "mapped_bucket_name": "test-list-objects",
+            "mapped_prefix": "dev/"
+        }
+
+    }
+    client_mapping = {"test_upload": minio1, "test_copy": minio1, "test_copy2": minio1,
+                      "test_delete_objects": minio2, "test_paginator": minio2, "test_list_objects": minio2,
+                      "default": minio2}
```
AFAICT this means that we test most actions only for being routed to one bucket.  So our testing looks like at most half of what it should be.

arielshaqed reviewed (commented):
I am very disappointed by this code review.  We went through over 100 comments -- that means that this code review is for too much code.  And we still have items marked "Done" that were not done.

A code review is _**NOT**_ a race to approval, it is a journey to improved understanding of the code by more people (represented by the reviewers), which improves code along the way.  If you agree to a comment but do not commit to making the change then you have frustrated the code review.  (You also slow down approval, in case you are still trying to race.)

I believe there is no further value in my participation as a reviewer of this code.  While it is generally good (and I approved it), by continuing to participate I will increase antagonism without advancing code.  @johnnyaug and @eden-ohana  -- sorry for bailing out right before the end.

eden-ohana commented:
@arielshaqed 

I am very disappointed to see the rough things that are written here.
As can be seen, some of the comments are marked as done and some are not, because I am still in the middle of the process! And that's why I haven't committed the code with the changes yet (I was planning to finish before my flight, as I agreed with yoni).
That's also why I haven't asked you to re-review it and haven't merged this PR yet.
What I’m doing is the complete opposite of racing- on the contrary, I’m fixing every little thing one by one and giving full attention to every comment submitted by you.

I feel that by writing this on Saturday you actually haven’t given me the chance to finish resolving all the comments.

I’m sorry to hear you are raising antagonism towards me even though I haven’t done anything wrong (or at least that I’m aware of), and of course haven’t done anything personal to you.

eden-ohana committed:
Code review changes

arielshaqed commented:
I apologise. I should not have insinuated. Sending the "done" comments on Wednesday without pushing commits was a mistake. Similarly it was an error for me to re-review. However on my review I ascribed motives. This [cognitive distortion](https://en.m.wikipedia.org/wiki/Cognitive_distortion) is similar to "mind reading (see that link). I know I am prone to falling into this mental trap, yet went through with it anyway.
